sycl/source/detail/kernel_impl.cpp

```diff
@@ -57,12 +57,32 @@ kernel_impl::~kernel_impl() {
   }
 }
 
+template <info::kernel param>
+typename info::param_traits<info::kernel, param>::return_type
+get_host_kernel_info() {
+  throw runtime_error(
+      "This kernel info parameter is not supported on HOST device.");
+}
+
+template <> string_class get_host_kernel_info<info::kernel::function_name>() {
+  return "Host kernel";
+}
+
+template <> cl_uint get_host_kernel_info<info::kernel::num_args>() { return 0; }
```

Comment:
This is not valid if any argument is passed to kernel.

Reply:
See the comment for attributes parameter.

Reply:
I think that the behavior like we have for  info::kernel::reference_count is more natural. 
We need to throw  invalid_object_error exception.

may be it is reasonable to update spec:
@bader to comment.







reference_count
--

Reply:
I think what @vladimirlaz is trying to highlight here is that users `0` can't be treated as valid number of arguments if kernel is compiled for the host device. 
@gmlueck, @jbrodman, @rolandschulz, @mkinsner, should we improve API here?
If I understand it correctly the result of this function can be used only if kernel compiled for non-host device, so it's user's responsibility to check this pre-condition before using `info::kernel::num_args` query result.

Reply:
This does not directly answer @bader's question, but I'll provide some more information.

The `info::kernel::num_args` descriptor still exists in SYCL 2020 provisional, and the spec specifically states that all kernel info descriptors must produce a valid value even for the host device:

> Every kernel (including a host kernel) must produce a valid
> value for each info parameter. The possible values for each info parameter and any restriction are defined in the
> specification of the SYCL backend associated with the kernel.

The description for `num_args` is a bit more generic in SYCL 2020 provisional:

> Return the number of arguments of the extracted kernel.

The `handler::set_arg()` interface also still exists in SYCL 2020 provisional:

> Set a kernel argument for a kernel through the SYCL backend interoperability interface.  The index value specifies which parameter of the low-level kernel is being set and arg specifies the kernel argument.  Index 0 is the first parameter.
> The argument can be either a SYCL accessor, a SYCL sampler or a trivially copyable C++ type.
> Note it is invalid to set arguments to SYCL kernel function objects.

I think the intention was to clarify the behavior and limitations of these APIs for the host backend in the "Host backend specification" appendix.  However, I do not see any such clarification there, so I think it was missed.

To be honest, I'm not sure what the right behavior should be, but I do not think we should throw an exception.  I think the two options are:

1. Say that `info::kernel::num_args` is not meaningful for the host platform and just return 0.
2. Emulate the same behavior as a real device and return the true number of parameters.

Obviously, the first is easier to implement, however, I've heard others state that the host device should emulate all (or nearly all) of the features of a real device.  Opinions of others listed above?

Reply:
> Emulate the same behavior as a real device and return the true number of parameters.

I guess it is not possible in a pure C++ host implementation while we do not have a lot of introspection landing in C++ (C++29?...).
To discuss in the SYCL standard committee what would be the best standard behaviour.

Reply:
As long as this compiler only promises SYCL 1.2.1 support, we can do whatever we wants for host-device because the 1.2.1 spec only requires the info to be valid for OCL kernels. I think throwing an exception is most sensible.

I agree with Ronan that we should clarify the SYCL 2020 spec regarding the expected behavior for the host platform.

Reply:
I created https://gitlab.khronos.org/sycl/Specification/-/issues/405 for this.